Use better algorithms

gasche opened the pull request:
(cc @nobrakal and @yurug)

Asak can be expensive to run on large corpus of code, so `src/clustering.ml` uses Parmap to parallelize computations. This PR proposes an alternative route, which is to improve the algorithms to go faster.

I propose a better algorithm for computing distances between all pairs of Lambda fingerprints, and a better algorithm to cluster the fingerprints into dendrograms.

I worked by first copying the file `src/clustering.ml` into a new file `src/clustering_bench.ml` which I then modified to measure runtime and experiment with algorithms. (I also removed all uses of Parmap and UnionFind in the code, and inlined the definition of Wtree; the file is completely self-contained). I wrote a random generator of fingerprints, that can be tuned to generate small or large sets of sub-hashes (large sets spend a lot of time in symmetric-difference copmutation), and small or large connected components (large components spend a lot of time in clustering). I edited the various parts one commit at a time, keeping the old code for performance comparison.

For example, running the program with fingerprint of cardinality 20 to generate 20 connected components of 150 fingerprints prints the following timing information:

```
[    0.024729] start
[   11.511494] compute_all_sym_diff done
[    0.809088] compute_all_sym_diff_fast done
[   24.843792] check compute_all_sym_diff
[    0.078135] convert_map_to_hm done
[    1.672870] surapproximate_classes done
[    1.925790] adjacency_list done
[    0.144706] surapproximate_classes_nouf done
[    0.005801] check surapproximate_classes
[   21.723186] hierarchical_clustering done
[    0.875388] hierarchical_clustering_fast done
```

On this run, the old algorithm for symmetric difference (after removing all the Parmap stuff, so purely sequential) takes 11.5s, and the new algorithm takes 0.8s. The old algorithm for clustering takes 21.7s, the new algorithm takes 0.9s. There is also a surapproximate_classes_nouf function (0.14s instead of 1.7s), for which the goal is not performance (this approximation is not a bottleneck of the previous code) but simplification: Union-Find is important for incremental unification computation (where you alternate same-connected-component queries with addition of new unification edges), but this is simply a connected-component computation that can be done with a graph traversal.

(The new algorithms are documented in code comments. At a high level: the symmetric-differences algorithm computes the distance from a node to all its neighbors in a single traversal of its own children, and the clustering algorithm uses a priority queue to avoid having to recompute the minimal pair at each step.)

For all_sym_diff and surapproximate_classes, there is a check that the new algorithm returns exactly the same result as the old algorithm. There is no such check for the clustering algorithm, because the results are *not* the same. There may be a bug in the new code, but there are also many reasons why two correct algorithms would return different results: the order of children in a dendrogram is irrelevant, and the algorithm does not specifies what to do when several pairs-of-trees are at the same minimal distance (typically, if all leaves are at the same distance from each other, any tree shape is a valid result). I tried to get the two algorithms to agree on their choices (by picking a minimal pair for lexicographic ordering), but it didn't work and it made the code more complex.

The old code represents distance information with a `('a * 'a, int) Hashtbl.t` structure. This makes it difficult to implement certain graph algorithms (typically, connected-components computation) where we need to get the list of all neighbors of a given node. For this reason I perform a conversion to an adjacency-list presentation in the middle of the code. I think that it would be simpler to use adjacency lists from the start, for example using a `('a, ('a, int) Hashtbl.t) Hashtbl.t` structure (my adjacency lists are simpler because I rely on the pair-hashmap for distance information). This would require a more global rewrite of the code, I didn't do it.

This PR changes only the code in `clustering_bench.ml`. If you like the algorithms proposed and wish to use them in production, they need to be integrated into `clustering.ml` proper. I didn't do it myself because I don't have the setup to reproduce the real-world experiments, or even conveniently play with the whole toolchain (so for me the isolated `clustering_bench` is easier).

gasche committed:
copy clustering.ml into a new file for modification

gasche committed:
clustering_bench: inline wtree.ml for self-containedness

gasche committed:
clustering_bench: remove Parmap support for self-containedness

gasche committed:
clustering_bench: fake input generation for benchmarking

gasche committed:
clustering_bench: show timings

gasche committed:
Distance: protect against non-normalized 'compare' results

gasche committed:
compute_all_sym_diff: faster algorithm

gasche committed:
surapproximate_classes: implementation without UnionFind

gasche committed:
fast implementation of the clustering algorithm

yurug commented:
Thank you Gabriel! That's super impressive.

nobrakal commented:
Thank you very much, this is indeed very impressive! I will try to read all this as soon as possible!

nobrakal commented:
Hi,

I finally had the time to read, understand and test your PR, excuse me for the latency.

The whole code runs _way_ faster than the actual slow version, even on big inputs (like all the hashes from `Opam`).

There is two (minor) issues:

* The set `was_seen` was badly named. I used it to remember elements that are not infinitely distant from everyone. This is not the case with the actual result of `compute_all_sym_diff_fast` (it takes into account too much elements), but can be simply recovered.
* A little more problematic, a parent can have multiple times the same child, and `compute_all_sym_diff_fast` seems to not take that into account when computing the symmetric difference. A simple way to overcome the problem is to add a suffix to distinguish same hashes.

Concerning the other part of the algorithm, the surapproximation and the clustering works fine on big examples (the fast and the slow algorithms return the exact same number of classes, so I think minor differences are not a problem)!

If you don't see any objections, I will merge this and adapt the code base to use your (super) fast algorithm :) . Again, many thanks !

gasche commented:
@nobrakal when I saw your comment on subterms with multiplicities (several occurrences), I started thinking about how to update the algorithm (tracking the number of times each subterm has been seen in the various datastructures). First I updated the test-data generator to make this case happen more often... and I didn't see any failures.

It is a surprise, but my impression is that the current implementation actually behaves correctly in this case: `Hashtbl.find_all` will return the list of all values that have been "added" to a key, and adding a value twice means that it will be present twice in the list. If you re-read the code while keeping this in mind, it sounds plausible (still surprising) that it works well for duplicated subterms.

I would recommend creating a concrete example testcase with repeated subterm, to ensure that this actually holds, but this is my current understanding.

Your general plan (to integrate the algorithms) seems fine. Thanks!

nobrakal commented:
Thanks for your response !

Sorry if I wasn't clear, here is a failing example (when both codes shares a common factor, but not with the same multiplicity):

```OCaml
let will_fail =
  let child = 1,"child" in
  let first = (100,"first"),[child] in
  let second = (100,"second"),[child; child] in
  [("first",first);("second",second)]
```

The symmetric difference here is just one `child`, and your algorithm does not seems to detect it (even with the `find_all` trick !).

Do you see any way to recover this instead of adding a different suffix to each "same" child ?

gasche committed:
a multi-occurrence version of the fast symmetric difference algorithm

In PR #1, nobrakal explained that the previous fast algorithm was
wrong when children are seen as a multiset rather than a set: distance
computation needs to take the number of occurrences of each child into
account.

gasche commented:
You were in fact clear, but I was too lazy to write a concrete testcase myself, so I tried to tweak my test generator, failed, and got the wrong conclusions.

I just added a new commit that adds a multi-occurrence version of the fast symmetric-difference algorithm. For reference (and benchmarking) I kept the old and wrong copy around. The new version works on your testcase (as the reference version); it is slightly slower than the previous fast algorithm (the data structures are slightly more complex, incurring more overhead) but is of the same complexity. 

Instead of tracking the set of parents that have a given children, I track for each child a map from its parents to its occurrence count in each parent (in your example, `"child"` maps to `[("first", 1); ("second", 2)]` as a hashtable). To compute the common weight between two parents, for each child, I take the minimum of the number of occurrences on both sides.

On a random test I have the following performance numbers:

```
[    2.182489] compute_all_sym_diff done
[    0.262365] compute_all_sym_diff_fast done
[    0.304869] compute_all_sym_diff_fast_multiocc done
```

The new `_multiocc` version takes 0.30s instead of 0.26s.

nobrakal commented:
Hi, thanks for your answer :)

The `_multiocc` version is still not working. Here is a slightly modified counter-example:

```Ocaml
let multiocc_testcase =
  (* from @nobrakal *)
  let child = 1,"child" in
  let first = (100,"first"),[child; child] in
  let second = (100,"second"),[child] in
  [("first",first);("second",second)]
```
As I understand it, the problem comes from that `add_child` (from `node_neighbors`) can be called with the same child multiple time, leading to an over-count of the common children (by chance my first counter-example used a good order which hide the bug)! 

I am not sure how to simply recover from this...

gasche committed:
multi-occurrence version: fix another counterexample of @nobrakal

gasche commented:
Ah, indeed. We need to iterate on each children without duplicates. I sent a small commit that turns `List.iter` into a conversion into a `HSet` first, followed by an iteration.

One approach to structure the code slightly differently would be to represent elements directly in this style, with a multi-set (typically as a `HMap` from children to occurrence count) instead of a list with duplicates. Then it would be harder to write my first incorrect version by mistake, and also slightly easier to compute the "transposed occurrence count" in the `parents` table.

gasche commented:
Thanks for the refined counter-example!

nobrakal commented:
@gasche Many thanks for your work ! The `_multiocc` version now behaves like the original one on the whole `Lambda` database I have.

I will merge this, adapt it to the actual code base and try the improvements you suggested :)
